sdk/servicebus/azure-servicebus/azure/servicebus/_common/utils.py

```diff
@@ -250,3 +250,44 @@ def shutdown(self, wait=True):
         :type wait: bool
         """
         self.executor.shutdown(wait=wait)
+
+
+class AutoComplete(object):
+    """
+    Provides a ContextManager to complete a message on successful exit from the scope,
+    abandoning() otherwise, for instance if an exception is thrown.
+
+    :param message: The message to auto-complete or abandon.
+    :type message: ~azure.servicebus.ReceivedMessage
+
+    .. admonition:: Example:
+
+        .. literalinclude:: ../samples/sync_samples/sample_code_servicebus_async.py
+            :start-after: [START auto_complete_sync]
+            :end-before: [END auto_complete_sync]
+            :language: python
+            :dedent: 4
+            :caption: Automatically complete a message
+    """
+    def __init__(self, message):
+        self.message = message
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if traceback:
+            try:
+                if not self.message._settled:
+                    _log.info("Attempting to auto-abandon message due to exception: " + str(value))
+                    self.message.abandon()
+            except (MessageAlreadySettled, MessageSettleFailed) as e:
+                _log.info("Unable to auto-abandon message: " + str(e))
+            return False
+        try:
+            if not self.message._settled:
+                _log.info("Attempting to auto-complete message")
+                self.message.complete()
+        except (MessageAlreadySettled, MessageSettleFailed) as e:
+            _log.info("Unable to auto-complete message: " + str(e))
```

Comment:
Dum question: would "MessageAlreadySettled, MessageSettleFailed" be raised out?
if it's not raised out, how would the user know that auto_complete failed to complete

Reply:
So you raise a good point, namely, should I ONLY catch MessageAlreadySettled.

The catch was initially so that if the user settled themselves, it would not spuriously alert them.  I'll make it more targeted unless you can think of any other exceptions we should be quashing. (to cover the scenario of "the user already took the needed action, either here or on another machine in parallel")

Reply:
catching only `MessageAlreadySettled` looks good to me.

sorry I don't fully understand "forgiving in terms of allowing double-settle", are you describing an non-autocomplete user scenario which does double-settle?

Reply:
Ah, sorry, to be more clear:  I was just second guessing the fact that I catch double-settle; towards the question of "if a user uses autocomplete, is that expressing enough intent that we should throw if they settle on their own?"

I don't THINK I agree with that statement (I think catching double settle makes this more flexible) But I was spitballing it to think out loud in case you had any realizations that I was missing.

Reply:
I got your question now..

In my understanding, auto-complete is a feature that "if user doesn't settle the message, we settle the message for them, if user settles, then we do nothing" which makes catching double-settle reasonable to me.